executor/coprocessor.go

```diff
@@ -0,0 +1,184 @@
+		err = Next(ctx, e, chk)
+		if err != nil {
+			break
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+		err = h.appendChunk(chk, tps)
+		if err != nil {
+			break
+		}
+	}
+	return h.buildResponse(err)
+}
+
+func (h *CoprocessorDAGHandler) buildDAGExecutor(req *coprocessor.Request) (Executor, error) {
+	if req.GetTp() != kv.ReqTypeDAG {
+		return nil, errors.Errorf("unsupported request type %d", req.GetTp())
+	}
+	dagReq := new(tipb.DAGRequest)
+	err := proto.Unmarshal(req.Data, dagReq)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	stmtCtx := h.sctx.GetSessionVars().StmtCtx
+	stmtCtx.SetFlagsFromPBFlag(dagReq.Flags)
+	stmtCtx.TimeZone, err = timeutil.ConstructTimeZone(dagReq.TimeZoneName, int(dagReq.TimeZoneOffset))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	h.dagReq = dagReq
+	return h.buildDAG(dagReq.Executors)
+}
+
+func (h *CoprocessorDAGHandler) buildDAG(executors []*tipb.Executor) (Executor, error) {
+	is := h.sctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
+	bp := core.NewPBPlanBuilder(h.sctx, is)
+	plan, err := bp.BuildPhysicalPlanFromPB(executors)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	b := newExecutorBuilder(h.sctx, is)
+	return b.build(plan), nil
+}
+
+func (h *CoprocessorDAGHandler) appendChunk(chk *chunk.Chunk, tps []*types.FieldType) error {
+	var err error
+	switch h.dagReq.EncodeType {
+	case tipb.EncodeType_TypeDefault:
+		err = h.encodeDefault(chk, tps)
+	case tipb.EncodeType_TypeChunk:
+		err = h.encodeChunk(chk, tps)
+	default:
+		return errors.Errorf("unknown dag encode type, %v", h.dagReq.EncodeType)
+	}
+	return err
+}
+
+func (h *CoprocessorDAGHandler) buildResponse(err error) *coprocessor.Response {
+	if err != nil {
+		h.resp.OtherError = err.Error()
+		return h.resp
+	}
+	h.selResp.EncodeType = h.dagReq.EncodeType
+	data, err := proto.Marshal(h.selResp)
+	if err != nil {
+		h.resp.OtherError = err.Error()
+		return h.resp
+	}
+	h.resp.Data = data
+	return h.resp
+}
+
+func (h *CoprocessorDAGHandler) encodeChunk(chk *chunk.Chunk, colTypes []*types.FieldType) error {
+	colOrdinal := h.dagReq.OutputOffsets
+	chunks := h.selResp.Chunks
+	respColTypes := make([]*types.FieldType, 0, len(colOrdinal))
+	for _, ordinal := range colOrdinal {
+		respColTypes = append(respColTypes, colTypes[ordinal])
+	}
+	encoder := chunk.NewCodec(respColTypes)
+	chunks = append(chunks, tipb.Chunk{})
+	cur := &chunks[len(chunks)-1]
+	cur.RowsData = append(cur.RowsData, encoder.Encode(chk)...)
+	h.selResp.Chunks = chunks
+	return nil
+}
+
+func (h *CoprocessorDAGHandler) encodeDefault(chk *chunk.Chunk, tps []*types.FieldType) error {
+	colOrdinal := h.dagReq.OutputOffsets
+	chunks := h.selResp.Chunks
+	requestedRow := make([]byte, 0)
+	for i := 0; i < chk.NumRows(); i++ {
+		requestedRow = requestedRow[:0]
+		row := chk.GetRow(i)
+		for _, ordinal := range colOrdinal {
+			data, err := codec.EncodeValue(h.sctx.GetSessionVars().StmtCtx, nil, row.GetDatum(int(ordinal), tps[ordinal]))
```

Comment:
It’s better to use local variable to stash the result of `h.sctx.GetSessionVars().StmtCtx` instead of call it every time in the loop.